aws/resource_aws_codebuild_source_credential.go

```diff
@@ -0,0 +1,130 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/codebuild"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/helper/validation"
+)
+
+func resourceAwsCodeBuildSourceCredential() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsCodeBuildSourceCredentialCreate,
+		Read:   resourceAwsCodeBuildSourceCredentialRead,
+		Delete: resourceAwsCodeBuildSourceCredentialDelete,
+
+		Schema: map[string]*schema.Schema{
+			"arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"auth_type": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					codebuild.AuthTypeBasicAuth,
+					codebuild.AuthTypePersonalAccessToken,
+				}, false),
+			},
+			"server_type": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					codebuild.ServerTypeGithub,
+					codebuild.ServerTypeBitbucket,
+					codebuild.ServerTypeGithubEnterprise,
+				}, false),
+			},
+			"token": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"user_name": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+func resourceAwsCodeBuildSourceCredentialCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).codebuildconn
+
+	authType := d.Get("auth_type").(string)
+
+	createOpts := &codebuild.ImportSourceCredentialsInput{
+		AuthType:   aws.String(authType),
+		ServerType: aws.String(d.Get("server_type").(string)),
+		Token:      aws.String(d.Get("token").(string)),
+	}
+
+	if attr, ok := d.GetOk("user_name"); ok && attr.(string) != "" && authType == codebuild.AuthTypeBasicAuth {
+		createOpts.Username = aws.String(attr.(string))
+	}
+
+	resp, err := conn.ImportSourceCredentials(createOpts)
+	if err != nil {
+		return fmt.Errorf("Error importing source credentials: %s", err)
+	}
+
+	d.SetId(aws.StringValue(resp.Arn))
+	d.Set("arn", resp.Arn)
+
+	return nil
+}
+
+func resourceAwsCodeBuildSourceCredentialRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).codebuildconn
+
+	resp, err := conn.ListSourceCredentials(&codebuild.ListSourceCredentialsInput{})
+	if err != nil {
+		return fmt.Errorf("Error list source credentials: %s", err)
+	}
+
+	if len(resp.SourceCredentialsInfos) == 0 {
+		log.Printf("[WARN] Source Credentials(%s) is already deleted", d.Id())
+		d.SetId("")
+		return nil
+	}
+
+	resourceNotFound := true
```

Comment:
Typically we prefer to use relevant object type as the output variable when searching like this, which simplifies the logic and allows the code to read well on the left side:

```go
var info *codebuild.SourceCredentialInfo

for _, sourceCredentialsInfo := range resp.SourceCredentialsInfos {
	if d.Id() == aws.StringValue(sourceCredentialsInfo.Arn) {
		info = sourceCredentialInfo
		break
	}
}

if info == nil {
	log.Printf("[WARN] CodeBuild Source Credential (%s) not found, removing from state", d.Id())
	d.SetId("")
	return nil
}

d.Set("arn", info.Arn)
d.Set("auth_type", info.AuthType)
d.Set("server_type", info.ServerType)

return nil
```

Reply:
fixed in 588a1c3